src/libkrun/src/lib.rs

```diff
@@ -484,14 +499,53 @@ pub unsafe extern "C" fn krun_set_mapped_volumes(
 #[allow(clippy::missing_safety_doc)]
 #[no_mangle]
 #[cfg(feature = "blk")]
-pub unsafe extern "C" fn krun_set_root_disk(ctx_id: u32, c_disk_path: *const c_char) -> i32 {
+pub unsafe extern "C" fn krun_add_disk(
+    ctx_id: u32,
+    c_block_id: *const c_char,
+    c_disk_path: *const c_char,
+    read_only: bool,
+) -> i32 {
     let disk_path = match CStr::from_ptr(c_disk_path).to_str() {
         Ok(disk) => disk,
         Err(_) => return -libc::EINVAL,
     };
 
-    //let fs_id = "/dev/root".to_string();
-    //let shared_dir = root_path.to_string();
+    let block_id = match CStr::from_ptr(c_block_id).to_str() {
+        Ok(block_id) => block_id,
+        Err(_) => return -libc::EINVAL,
+    };
+
+    // These block IDs are special cased in the block_cfgs logic for legacy compatibility. Just ban
+    // them for the new API, force the user to avoid unclear/ambiguous API usage.
+    if block_id == "root" || block_id == "data" {
```

Comment:
Instead of banning them outright, I *think* you could do something like:

`krun_set_root_disk` would forward to `krun_add_disk` with `c_block_id = "root"`
`krun_set_data_disk` would forward to `krun_add_disk` with `c_block_id = "data"`

`krun_add_disk` would have a special case with `c_block_id = "root"` to set the root disk.

AFAIU, there would be no functional difference between any added disk *except* the root disk, correct? So `krun_set_data_disk` may be able to just forward to `krun_add_disk`.

Just trying to make sure we don't end up with useless APIs when replacements would be preferred.

Reply:
> AFAIU, there would be no functional difference between any added disk except the root disk, correct? So krun_set_data_disk may be able to just forward to krun_add_disk.

Even the root disk is just a data disk here, there's no special cases whatsoever.

The problem is backwards compat. It's tempting to redefine set_*_disk as adds. However, consider the follow application code:

```
krun_set_data(foo)
krun_set_data(bar)
krun_set_root(baz)
```

Currently, that results in exactly two partitions: root=baz and data=bar, in that guaranteed order.

If we aliased the sets to adds, however, we get three partitions `data=foo, data=bar, root=baz`, in the wrong order, and with colliding block IDs.

If we didn't have a stable API I'd just delete the root/data methods altogether. The compromise I added here is carefully written to ensure no change to existing users. Restricting just the IDs "root/""data" makes it very easy to do that without introducing piles of logic with unclear semantics. (if you add "root" and then set_root, what happens? what if you set_root but then add "root"? etc. Lot of hairy interactions without clear answers. Better to just ban it and do the simplest correct thing.)

Reply:
But for the example you just gave:
```
krun_set_data(foo)
krun_set_data(bar)
krun_set_root(baz)
```
I don't see why that application would exist in the first place. Specifically because a user would not set the data disk twice, since the second call would essentially replace the data disk set first, as you pointed out with `root=baz, data=bar`.

Reply:
I think there's no "right" solution here, either approach would be fine by me. We'll fix this mess (and many others in the API) in 2.0.

Reply:
> I don't see why that application code would (currently) exist in the first place

I mean, sure. But I'm not about to break b/w compat without a semver major bump.

My recommendation remains thus: have the gunk here to retain old behaviour, consider the set* methods deprecated, and then in 2.0 remove the `set*` methods in favour of only `add` and then you can garbage collect this legacy cruft.